lib/Target/OpenFhePke/Interpreter.cpp

```diff
@@ -137,9 +112,268 @@ using PrivateKeyT = PrivateKey<DCRTPoly>;
+          doubleValues.erase(v);
+        }
+      })
+      .Case<openfhe::DigitDecompositionType>(
+          [&](auto ty) { fastRotPrecomps.erase(v); })
+      .Case<openfhe::EvalKeyType>([&](auto ty) { evalKeys.erase(v); })
+      .Case<openfhe::PublicKeyType>([&](auto ty) { publicKeys.erase(v); })
+      .Case<openfhe::PrivateKeyType>([&](auto ty) { privateKeys.erase(v); })
+      .Case<openfhe::CryptoContextType>(
+          [&](auto ty) { cryptoContexts.erase(v); })
+      .Default([&](Type type) {
+        llvm::errs() << "Unsupported type " << type << " in eraseValue\n";
+      });
+}
+
+// Only used for type-agnostic block arguments (func args, iter args, etc.)
+void Interpreter::storeTypedValue(Value v, const TypedCppValue& typedVal) {
+  std::visit(
+      [&](auto&& arg) {
+        using T = std::decay_t<decltype(arg)>;
+        if constexpr (std::is_same_v<T, std::monostate>) {
+          // Nothing to store
+        } else if constexpr (std::is_same_v<T, bool>) {
+          boolValues[v] = arg;
+        } else if constexpr (std::is_same_v<T, int>) {
+          intValues[v] = arg;
+        } else if constexpr (std::is_same_v<T, float>) {
+          floatValues[v] = arg;
+        } else if constexpr (std::is_same_v<T, double>) {
+          doubleValues[v] = arg;
+        } else if constexpr (std::is_same_v<
+                                 T, std::shared_ptr<std::vector<int>>>) {
+          intVectors[v] = arg;
+        } else if constexpr (std::is_same_v<
+                                 T, std::shared_ptr<std::vector<float>>>) {
+          floatVectors[v] = arg;
+        } else if constexpr (std::is_same_v<
+                                 T, std::shared_ptr<std::vector<double>>>) {
+          doubleVectors[v] = arg;
+        } else if constexpr (std::is_same_v<T, PlaintextT>) {
+          plaintexts[v] = arg;
+        } else if constexpr (std::is_same_v<
+                                 T, std::shared_ptr<std::vector<PlaintextT>>>) {
+          plaintextVectors[v] = arg;
+        } else if constexpr (std::is_same_v<T, CiphertextT>) {
+          ciphertexts[v] = arg;
+        } else if constexpr (std::is_same_v<T, std::shared_ptr<
+                                                   std::vector<CiphertextT>>>) {
+          ciphertextVectors[v] = arg;
+        } else if constexpr (std::is_same_v<T, PublicKeyT>) {
+          publicKeys[v] = arg;
+        } else if constexpr (std::is_same_v<T, PrivateKeyT>) {
+          privateKeys[v] = arg;
+        } else if constexpr (std::is_same_v<T, EvalKeyT>) {
+          evalKeys[v] = arg;
+        } else if constexpr (std::is_same_v<T, CryptoContextT>) {
+          cryptoContexts[v] = arg;
+        } else if constexpr (std::is_same_v<T, FastRotPrecompT>) {
+          fastRotPrecomps[v] = arg;
+        }
+      },
+      typedVal.value);
+}
+
+// Only used for type-agnostic block terminators (return, yield)
+TypedCppValue Interpreter::loadTypedValue(Value v) {
+  // Try each storage type
```

Comment:
it seems a little rough to search through each storage type - and in `eraseValue` there is also a switch on MLIR type -> storage map. Could there by a map from MLIR type to a pointer to the map that you could use so you can do something like `mlirTypeToContainer[v.getType()]->at(v)` and then in `eraseValue` it can also to `mlirTypeToContainer[v.getType()]->erase(v)`?
Or is that indirection more overhead?

Reply:
It only shows up as a cumulative 10ms in the profile (10ms for loadTypedValue, and 10ms for eraseValue, a total of 0.02%), which I think is sufficiently small to ignore.